NeoML/src/TraditionalML/KMeansClustering.cpp

```diff
@@ -55,9 +55,208 @@ static CPtr<CDnnBlob> createWeightBlob( IMathEngine& mathEngine, const IClusteri
+		for( int c = 0; c < owner.clusters.Size(); ++c ) {
+			if( isPruned( c, i ) ) {
+				continue;
+			}
+			float dist = UpperBounds[i];
+			if( mustRecalculate ) {
+				mustRecalculate = false;
+				auto& cluster = *owner.clusters[Assignments[i]];
+				auto rowDesc = Matrix->GetRow( i );
+				dist = static_cast<float>( sqrt( cluster.CalcDistance( rowDesc, distanceFunc ) ) );
+				LowerBounds( Assignments[i], i ) = dist;
+				UpperBounds[i] = dist;
+			}
+			if( dist > LowerBounds( c, i ) || dist > 0.5 * ClusterDists( Assignments[i], c ) ) {
+				auto& cluster = *owner.clusters[c];
+				auto rowDesc = Matrix->GetRow( i );
+				const float pointDist = static_cast<float>( sqrt( cluster.CalcDistance( rowDesc, distanceFunc ) ) );
+				LowerBounds( c, i ) = pointDist;
+				if( pointDist < dist ) {
+					UpperBounds[i] = pointDist;
+					Assignments[i] = c;
+				}
+			}
+		}
+	}
+}
+
+//-------------------------------------------------------------------------------------------------------------
+
+struct CKMeansClustering::CUpdateULBoundsArgs final : public CKMeansClustering::IArgs {
+	CUpdateULBoundsArgs( const CFloatMatrixDesc& matrix, CAssignVectorsArgs& assigns, int threadCount ) :
+		IArgs( &matrix ),
+		Assigns( assigns ),
+		Inertia( threadCount )
+	{}
+
+	void Run( const CKMeansClustering&, int threadIndex, int operation ) override;
+
+	CAssignVectorsArgs& Assigns;
+	CInertia<16> Inertia;
+};
+
+void CKMeansClustering::CUpdateULBoundsArgs::Run( const CKMeansClustering& owner, int threadIndex, int j )
+{
+	for( int c = 0; c < owner.clusters.Size(); ++c ) {
+		Assigns.LowerBounds( c, j ) = max( Assigns.LowerBounds( c, j ) - Assigns.MoveDistance[c], 0.f );
+	}
+	Assigns.UpperBounds[j] += Assigns.MoveDistance[Assigns.Assignments[j]];
+	Inertia.Get( threadIndex ) += owner.clusters[Assigns.Assignments[j]]->CalcDistance( Matrix->GetRow( j ), owner.params.DistanceFunc );
+}
+
+//-------------------------------------------------------------------------------------------------------------
+
+struct CKMeansClustering::CVectorCopyArgs final : public CKMeansClustering::IArgs {
+	CVectorCopyArgs( int vectorSize, IMathEngine &mathEngine,
+			const CFloatHandle& firstHandle, const CConstFloatHandle& secondHandle ) :
+		IArgs( vectorSize ),
+		MathEngine( mathEngine ),
+		FirstHandle( firstHandle ),
+		SecondHandle( secondHandle )
+	{}
+
+	void Run( const CKMeansClustering&, int index, int count ) override
+	{ MathEngine.VectorCopy( FirstHandle + index, SecondHandle + index, count ); }
+
+	IMathEngine& MathEngine;
+	const CFloatHandle& FirstHandle;
+	const CConstFloatHandle& SecondHandle;
+};
+
+//-------------------------------------------------------------------------------------------------------------
+
+struct CKMeansClustering::CThreadTask final {
+	CThreadTask( const CKMeansClustering& owner, IArgs& args ) :
+		owner( owner ),
+		args( args )
+	{}
+
+	void CallRun( int threadIndex );
+	void CallRun( int threadIndex, int align );
```

Comment:
IMHO, you don't need two functions, add default value `1` for the `align` parameter.

Reply:
Because different implementation of methods

Reply:
I agree with @lordnn, better to have 1 method and and have `IArgs::Run` which accepts 4 arguments: `clustering, threadIndex, firstElem, elemCount` instead of treating differently 2 integer arguments from the `IArgs` in different realizations.
Now you have `IArgs::Run( clustering, threadIndex, ??operation?? )` and `CVectorCopyArgs::Run( clustering, firstElemIndex, elemCount )`

Reply:
Or even better not `elemIndex, elemCount` burt `vectorIndex, vectorCount` copyVector just multiplies `vectorCount` by vector size.